[WIP] wrap XLATensors into Python XLA wrapper class

Krovatkin opened the pull request:
**Improbably** fixes https://github.com/pytorch/pytorch/issues/65130

@ezyang I'm super n00b in Python extensions, is this what we want to do?

Krovatkin committed:
wrap XLATensors into Python XLA wrapper class

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -199,6 +206,13 @@ PyObject * THPVariable_Wrap(at::TensorBase var)
   }
+
+  // TODO: is there a better check?
+  if (var.device().type() == c10::kXLA) {
```
I suppose you could check the dispatch key, but device is pretty reasonable

albanD replied:
Also `var.is_xla()` does already exist :)

Krovatkin commented:
fyi:  
@miladm

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -154,6 +154,13 @@ static bool check_has_torch_dispatch(PyObject *obj) {
+  // TODO: What should we do if XLATensor doesn't exist? Return THPVariableClass?
+  // TODO: is THPObjectPtr the right pointer container?
+  static THPObjectPtr xla_tensor_ptr(py::globals()["XLATensor"].ptr());
+  return xla_tensor_ptr.get();
```
It's better to leak the PyObject as the destructor will run at shutdown when the Python interpreter may have already exited. See also how THPVariableClass is implemented.

Note that unlike THPVariableClass we cannot assume that the XLATensor is available at PyTorch time, since it doesn't show up until XLA loads itself. So it might be more appropriate to have some private C api for "Registering" the class, and then having XLA register the class itself at load time

Krovatkin replied:
> Note that unlike THPVariableClass we cannot assume that the XLATensor is available at PyTorch time, since it doesn't show up until XLA loads itself. 

I was thinking that the XLA extension is loaded before we could start creating XLA tensors, so hopefully, we can work with XLA guys to make sure `XLATensor` appears in `globals()` before we would try to create XLA tensors?

Also, I wonder if it would be acceptable to fallback to `THPVariableClass` if `XLATensor` isn't available? 

> So it might be more appropriate to have some private C api for "Registering" the class, and then having XLA register the class itself at load time

hehe, this sounds like a proper solution you mentioned @albanD might be working on :) I could work with him if we can't figure out an acceptable hack.

ezyang replied:
> I was thinking that the XLA extension is loaded before we could start creating XLA tensors, so hopefully, we can work with XLA guys to make sure XLATensor appears in globals() before we would try to create XLA tensors?

I guess these are both valid ways to skin the cat. But it feels weird for PyTorch to reach out to a module that doesn't exist anywhere in PyTorch: you call me versus I'll call you.

Krovatkin replied:
fair enough! I put together another quick sketch. I wonder what you and @albanD  think

albanD commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -154,6 +154,13 @@ static bool check_has_torch_dispatch(PyObject *obj) {
+static PyObject* getXLATensorClass() {
+  // TODO: What should we do if XLATensor doesn't exist? Return THPVariableClass?
+  // TODO: is THPObjectPtr the right pointer container?
+  static THPObjectPtr xla_tensor_ptr(py::globals()["XLATensor"].ptr());
```
I'm not familiar with the pybind globals API but how is the error raised when the key is not there? Should you handle it like a C-python error? Or is it a c++ error?

Krovatkin replied:
@albanD  We could check if the key exists.
```
static PyObject* xla_tensor_python_class = py::globals().contains("XLATensor) ? py::globals()["XLATensor"].ptr() : nullptr;
return xla_tensor_python_class ;
```

**I wonder what should we do if the key doesn't exist?** Would it be acceptable to fall back to THPVariableClass? cc: @ezyang or should we throw?

albanD replied:
When would that happen? If the users loads an older version of torch_xla that doesn't have that class?

Krovatkin replied:
> If the users loads an older version of torch_xla that doesn't have that class?

Yeah, I'm not sure this is very likely, so maybe we shouldn't worry about being a backward compatible with older XLA versions and just throw? 

> When would that happen? 

A great q. I don't know really 😸  the XLA extension was partially initialized, so we somehow can create XLA cpp tensors, but some python classes are missing.

albanD commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -154,6 +154,13 @@ static bool check_has_torch_dispatch(PyObject *obj) {
+  // TODO: What should we do if XLATensor doesn't exist? Return THPVariableClass?
+  // TODO: is THPObjectPtr the right pointer container?
+  static THPObjectPtr xla_tensor_ptr(py::globals()["XLATensor"].ptr());
+  return xla_tensor_ptr.get();
```
You want to use release here otherwise xla_tensor_ptr might get de-allocated? (even though is is super unlikely as it is a type)
And the caller is then responsible for getting rid of that reference.

If you want to go yolo and just no do any refcount at all then you should use plain PyObjects here.

Krovatkin replied:
hehe, @ezyang says we should go yolo, we are doing that with `THPVariableClass` anyways, it seems  😺

albanD replied:
The tp_alloc/tp_dealloc do the proper incref/decref.
So the only problem thst can happen is that another thread frees the type between here and the newWithVar call below. I agree that this is unlikely enough :p

Krovatkin committed:
address initial feedback

Krovatkin committed:
add registration mechanism

Krovatkin commented:
@ezyang  @albanD how about this approach. Note, this is pretty much a sketch at this point: all I know that it compiles. I would like to get some validation before moving forward.

wconstab commented:
I like this approach (the registration API you expose, and having it called from the LTC backends during their init).

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -154,6 +154,20 @@ static bool check_has_torch_dispatch(PyObject *obj) {
   );
 }
 
+static std::unordered_map<c10::Device, PyObject*> device_to_py_ten_class_;
```
This is gonna be a major perf regression

ezyang commented:
The API sketched seems fine but (1) you need to speed it up (personally, my goal would be at most one easily predicted condition branch on PyObject tensor hotpath) and (2) it should be illegal for users to override wrapper class for builtins like CPU/CUDA, where this API provides unwanted public API surface.

Krovatkin committed:
address Ed's feedback

Krovatkin commented:
@ezyang Tried to address both points: 

> (1) you need to speed it up (personally, my goal would be at most one easily predicted condition branch on PyObject tensor hotpath) 


Added a check for CPU, CUDA before we even reach for a mapping. Make a mapping a regular array indexed by `DeviceType` Alternatively, we could always use the mapping by pre-populating it for built-in types.

> and (2) it should be illegal for users to override wrapper class for builtins like CPU/CUDA, where this API provides unwanted public API surface.

Added a check to forbid overriding CPU, CUDA.

zou3519 commented on torch/csrc/autograd/python_variable.h:
```diff
@@ -18,6 +18,8 @@ struct THPVariable {
   PyObject* backward_hooks = nullptr;
 };
 
+TORCH_API void RegisterPythonTensorClass(std::string device, PyObject* python_tensor_class);
```
nit: `const std::string&`, otherwise std::string gets passed by value

zou3519 commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -154,6 +155,22 @@ static bool check_has_torch_dispatch(PyObject *obj) {
+void RegisterPythonTensorClass(std::string device, PyObject* python_tensor_class) {
+  c10::Device dev(device);
+  if (dev.type() == kCPU || dev.type() == kCUDA) {
+    // TODO: should I throw something else?
+    throw new std::runtime_error("Cannot override classes for kCPU or kCUDA");
```
Does TORCH_CHECK work here?

zou3519 commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -154,6 +155,22 @@ static bool check_has_torch_dispatch(PyObject *obj) {
+    // TODO: should I throw something else?
+    throw new std::runtime_error("Cannot override classes for kCPU or kCUDA");
+  }
+  device_to_py_class_[static_cast<size_t>(dev.type())] = python_tensor_class;
```
It might be good to check that someone didn't register something previously or warn on it?

zou3519 commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -154,6 +155,22 @@ static bool check_has_torch_dispatch(PyObject *obj) {
 }
 
+
+static PyObject* device_to_py_class_ [static_cast<size_t>(c10::DeviceType::COMPILE_TIME_MAX_DEVICE_TYPES)];
```
I forget, are arrays zero-initialized?

zou3519 reviewed (commented):
seems pretty reasonable to me

ezyang commented:
> Added a check for CPU, CUDA before we even reach for a mapping. Make a mapping a regular array indexed by DeviceType Alternatively, we could always use the mapping by pre-populating it for built-in types.

I feel like I want to go even further: can we ONLY allow XLA device to be overwritten in this API? (You can add LTC too, if you like) While I am not opposed to other non-core types using this API, I want to know when someone starts using this API, and the easiest way to make sure that happens is to force them to send us a PR to add an extra exclusion for their use case.

Krovatkin commented on torch/autograd/variable.py:
```diff
@@ -11,6 +11,12 @@ def __instancecheck__(cls, other):
+        raise RuntimeError("clazz isn't a typeinfo object")
+    # TODO maybe we should have a check for CPU, CUDA here
+    # rather than buried deep in `RegisterPythonTensorClass`
+    torch._C._autograd._register_py_class_for_device
```
yes, my bad, looks like I didn't push the whole thing :(

Krovatkin committed:
clean up py files

Krovatkin committed:
address Ed's feedback

ezyang reviewed (approved):
sure why not

albanD reviewed (approved):
LGTM
